File: Arrays/zeros-to-end/zerosToEnd.js
function zerosToEnd(arr) {
  let writeIndex = 0;

  for (let readIndex = 0; readIndex < arr.length; readIndex++) {
    if (arr[readIndex] !== 0) {
      arr[writeIndex] = arr[readIndex];
      writeIndex++;
    }
  }

  while (writeIndex < arr.length) {
    arr[writeIndex] = 0;
    writeIndex++;
  }
}

const array = [0, 1, 0, 2, 3, 0, 4, 0, 5];
zerosToEnd(array);
console.log(array);

File: Arrays/zeros-to-end/zerosToEnd.go
package main

import "fmt"

func zerosToEnd(arr []int) []int {
	if len(arr) == 0 {
		return arr
	}

	writeIndex := 0
	for readIndex := 0; readIndex < len(arr); readIndex++ {
		if arr[readIndex] != 0 {
			arr[writeIndex] = arr[readIndex]
			writeIndex++
		}
	}

	for writeIndex < len(arr) {
		arr[writeIndex] = 0
		writeIndex++
	}
	return arr
}

func main() {
	array := []int{0, 1, 2, 0, 3, 4}
	array = zerosToEnd(array)
	fmt.Println(array)
}
